pytorch_lightning/trainer/supporters.py

```diff
@@ -120,26 +120,106 @@ def mean(self):
 
 
 class PredictionCollection(object):
-    def __init__(self, global_rank: int, world_size: int):
+
+    ID_KEY = 'id'
+
+    def __init__(self, global_rank: int, world_size: int, all_gather_fn: Callable):
         self.global_rank = global_rank
         self.world_size = world_size
-        self.predictions = {}
-        self.num_predictions = 0
+        self.all_gather_fn = all_gather_fn
+        self._legacy_predictions = {}
+        self._predictions = {stage: {} for stage in LoggerStages}
+
+    @property
+    def predictions(self) -> Dict:
+        return self._predictions[self.current_stage]
+
+    def _cache_prediction(self, predictions: List[Dict], dl_idx: int, batch_indices: List[int]) -> None:
+        cache = self.predictions
+
+        cache.setdefault(dl_idx, {})
+
+        def convert(value):
+            return value.cpu().tolist()
+
+        for batch_index, pred in zip(batch_indices, predictions):
+            if batch_index in cache[dl_idx]:
+                raise MisconfigurationException(
+                    "Prediction Collection doesn't support multiple predictions for one sample yet.")
+
+            cache[dl_idx][batch_index] = {
+                self.ID_KEY: batch_index, "predictions": apply_to_collection(pred, torch.Tensor, convert)}
```

Comment:
I assume we're converting to torch tensors for the all gather op?

Reply:
We are converting them to save gpu memory, and then I should improve the all_gather logic, so it moves from cpu -> gpu -> cpu per sample and not everything a